Реализовать задачу 5 спринта

glomebom opened the pull request:
Реализовать задачу 5 спринта

glomebom committed:
Разделить ответственность

glomebom committed:
Реализовать делегата

glomebom committed:
Исправить ошибки делегата

glomebom committed:
Исправить ошибки делегата
Убрать опционал фабрики

glomebom committed:
Спринт 5/20: 5 → Тема 2/4: Ответственность → Урок 9/9

В структуре AlertModel должны быть:
    + текст заголовка алерта title,
    + текст сообщения алерта message,
    + текст для кнопки алерта buttonText,
    - замыкание без параметров для действия по кнопке алерта completion.

Контроллер в методе окончания игры должен:
    + создавать модель для AlertPresenter,
    + передавать её в написанный метод этого класса для отображения алерта.

glomebom committed:
Выполнить задание: Спринт 5/20: 5 → Тема 3/4: Хранение данных → Урок 7/10

glomebom committed:
Реализовать задачу 5 спринта

Nikitaaeee commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+class AlertPresenter {
+    
```
Для улучшения производительности можно добавлять модификатор `final` к классам, от которых мы не будем наследоваться. (Как у тебя сделано в **MovieQuizViewController**, например) 
В таком случае комплятор будет быстрее находить адрес ячейки памяти при вызове функции.💡

Nikitaaeee commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,67 @@
+//
+
+import Foundation
+
+class QuestionFactory: QuestionFactoryProtocol {
+    
```
Этому классу тоже можно добавить **final**)

Nikitaaeee commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,67 @@
+        }
+        
+        let question = questions[safe: index]
+        delegate?.didReceiveNextQuestion(question: question)
+        
+    }
```
Пустые строки в конце методров/классов принято удалять, SwiftLint (сервис для чистого кода) может ругаться на такое (оставлять warning'и) 👀

Nikitaaeee commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,139 @@
+    var totalAmount: Int { get set }
+    func store(correct count: Int, total amount: Int)
+}
+
+class StatisticServiceImplementation: StatisticService {
```
Тут тоже можно добавить **final**.
Я обычно добавляю **final** каждый раз при создании нового класса, а если в дальнейшем надо будет от него наследоваться, то компилятор подскажет, что проблема в `final` и его можно будет убрать :)

glomebom committed:
Исправить:
- Класс StatisticServiceImplementation закрыт протоколом StatisticService. В MovieQuizViewController используется переменная statisticService типа StatisticService.
- Доступность кнопок в момент паузы перед показом следующего вопроса.
- Префикс final к классам наследование которых не предусмотрено.
- Пустые строки в конце методов/классов.
